### ISSUE TITLE
MemoryError when calling fit() for MapieTimeSeriesRegressor

### ISSUE THREAD

**Issue by silex-thomas** [scikit-learn-contrib/MAPIE#326]
**Describe the bug**
It seems that MAPIE is creating a matrix of size `n_training_samples*n_training_samples` when calling fit() on MapieTimeSeriesRegressor. Is this really the intended behaviour? That seems highly impractical as 50k samples for example is hardly considered big for a dataset. If so, is there a workaround?

**Minimal Reproducible Example**
the following MRE:
```
from sklearn.datasets import make_regression 
from mapie.subsample import BlockBootstrap
from mapie.regression import MapieTimeSeriesRegressor

X, y = make_regression(n_samples=50000, n_features=50, noise=20, random_state=59)
dummy_model = Ridge(alpha=0.05)
dummy_model.fit(X, y)


cv_mapiets = BlockBootstrap(n_resamplings=5, length=5000, overlapping=True, random_state=59)
mapie_enbpi = MapieTimeSeriesRegressor(dummy_model, method="enbpi", cv=cv_mapiets, agg_function="mean", n_jobs=1)
mapie_enbpi.fit(X, y)
```
 outputs for me: `MemoryError: Unable to allocate 18.6 GiB for an array with shape (50000, 50000) and data type float64`

Thank you for your help.

**Comment by thibaultcordier** [comment]
Hello @silex-thomas, thank you for reporting the bug you encountered.

### Bug Description

TL;DR: This is a problem that can arise when the calibration set has a larger number of samples. This behavior is unintended, as the `predict` method, called in the `fit` method of `MapieTimeSeriesRegressor`, is generally used with a smaller number of test samples during inference.

When calling `fit(X_calib, y_calib)` on `MapieTimeSeriesRegressor`, MAPIE calls `super().fit(X_calib, y_calib)` and `super().predict(X_calib)`. The `predict` call creates a matrix of size `(n_calib_samples, n_calib_samples)`.

Concerning the behaviour of the `predict` method, as explained in the [theoretical description of the documentation](https://mapie.readthedocs.io/en/latest/theoretical_description_regression.html) (see Fig of CV+), MAPIE needs to compute the distribution of residuals and predictions for all calibration samples for each test point. MAPIE does it in a vectorized way with two-dimensional arrays of size `(n_calib_samples, n_test_samples)`, hence exceeding the memory when the number of calibration and test samples are high.

### Short- and medium-term solution 

In the short term, I don't have a satisfactory workaround. One possibility is to reduce the number of samples to 10,000 (you can go up to 20,000, but the calculation time becomes very long). A more satisfactory solution would be to use the parameter `cv="prefit"` to separate training from calibration:

```python
import numpy as np
from sklearn.datasets import make_regression
from sklearn.linear_model import LinearRegression 
from mapie.regression import MapieTimeSeriesRegressor

X_train, y_train = make_regression(n_samples=50000, n_features=50, noise=20, random_state=59)
X_calib, y_calib = make_regression(n_samples=10000, n_features=50, noise=20, random_state=59)
X_test, y_test = make_regression(n_samples=1000, n_features=50, noise=20, random_state=59)

dummy_model = LinearRegression()
dummy_model.fit(X_train, y_train)
print('Training done')

mapie_enbpi = MapieTimeSeriesRegressor(
    dummy_model, method="enbpi", cv="prefit", agg_function="mean"
)
mapie_enbpi.fit(X_calib, y_calib)
print('Calibration done')

alpha = 0.05
gap = 10
y_pred_npfit, y_pis_npfit = mapie_enbpi.predict(X_calib, alpha=alpha)
y_pred_pfit, y_pis_pfit = np.zeros(y_pred_npfit.shape), np.zeros(y_pis_npfit.shape)
y_pred_pfit[:gap], y_pis_pfit[:gap, :, :] = mapie_enbpi.predict(
    X_test[:gap, :], alpha=alpha, ensemble=True, optimize_beta=True
)
for step in range(gap, len(X_test), gap):
    mapie_enbpi.partial_fit(
        X_test[(step - gap):step, :], y_test[(step - gap):step],
    )
    (
        y_pred_pfit[step:step + gap],
        y_pis_pfit[step:step + gap, :, :],
    ) = mapie_enbpi.predict(
        X_test[step:(step + gap), :], alpha=alpha, ensemble=True, optimize_beta=True
    )
print('Prediction done')
```

In the medium term, we can take a closer look and modify the code to reduce the calculation cost.

**Comment by silex-thomas** [comment]
hi @thibaultcordier, thanks a lot for such a detailed answer. I will reduce the number of samples as you suggest.
